### PR TITLE
gh-88773: Added teleport method to Turtle library

### LINKED ISSUE #88773: Teleport method for turtle class
BPO | [44607](https://bugs.python.org/issue44607)
--- | :---
Nosy | @serhiy-storchaka, @vedgar, @begilbert-sys

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-07-12.06:48:56.540>
labels = ['type-feature', 'expert-tkinter', '3.11']
title = 'Teleport method for turtle class'
updated_at = <Date 2021-07-12.17:42:12.710>
user = 'https://github.com/begilbert-sys'
```

bugs.python.org fields:
```python
activity = <Date 2021-07-12.17:42:12.710>
actor = 'Muffinlicious'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Tkinter']
creation = <Date 2021-07-12.06:48:56.540>
creator = 'Muffinlicious'
dependencies = []
files = []
hgrepos = []
issue_num = 44607
keywords = []
message_count = 4.0
messages = ['397286', '397292', '397303', '397341']
nosy_count = 3.0
nosy_names = ['serhiy.storchaka', 'veky', 'Muffinlicious']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue44607'
versions = ['Python 3.11']
```

</p></details>

### Linked PRs
* gh-103974

### PR DESCRIPTION
**Pull request by liam-gersten**
Fully implemented and tested a teleport method as described by the issue. This method behaves differently from the `goto` (or `setpos`) method, and can be especially helpful when creating and filling multiple separate objects. 

Implementation:

There already exists a `goto` method, aliased as `setpos | setposition | goto`. It basically moves the turtle to the given position, but if the pen is down, it draws a line along the path. Additionally, the move is not instant, and you can see it go along the path to get to the destination. Users of the library may find themselves raising the pen, calling the method, and then lowering it so that a line doesn’t get drawn. Another issue with using goto for this is that regardless of whether the pen is raised, an "imaginary" line is still drawn from the current position to the destination. When filling is enabled, this invisible line acts as a boundary for the fill color. If one were creating multiple separate objects with filling enabled, they would need to run the exact code in the teleport method in order to fill them separately as desired. The following is a [demo](https://github.com/liam-gersten/cpython/blob/e425d1cef7429fb7863f8d24af1215934ddc5e77/project_demo.py) of the difference between `goto` and `teleport` for filling multiple objects.

Justification:

One might think we should just update the `goto` method to behave this way with a flag like `teleport=True`. However, all of the pen controlling methods are in one class, `TPen`, and all of the visual and position controlling methods are in another, `TNavigator`. The primary turtle class, `RawTurtle(TPen, TNavigator)` inherits methods from both of these, and since teleport needs access to methods and attributes in both of these classes, this is where it should be written. While it would be possible to give `goto` access to the `TPen` class via inheritance, this would be a messy refactor for what should be a distinct method.

Resolves https://github.com/python/cpython/issues/88773

* Issue: gh-88773

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2023-04-28-18-04-23.gh-issue-88773.xXCNJw.rst
Comment (terryjreedy) [inline Misc/NEWS.d/next/Library/2023-04-28-18-04-23.gh-issue-88773.xXCNJw.rst]:
```suggestion
Added :func:`turtle.teleport` to turtle library to move a turtle to a new point without tracing a line, visible or invisible.  Patch by Liam Gersten.
```

File: Lib/turtle.py
Comment (gpshead) [inline Lib/turtle.py]: These 7 lines could be written as two using conditional expressions:

```
new_x = x if x is not None else self._position[0]
new_y = y if y is not None else self._position[1]
```

**Review by terryjreedy — CHANGES_REQUESTED** [review]
I cannot comment on the desirability of the addition until I have time to test and verify justification assertion.  Not now.
Nor have I reviewed the new tkinter code.

Blurb needs expansion.  I will commit my proposal and you can propose further revision.

Doc and test additions are needed before merge.

Note: Topics are for issues.  Core developers get notified of PRs by signing up as 'code owner' of specific files.  Besides which, this does not touch tkinter.

**Review by gpshead — CHANGES_REQUESTED** [review]
Please also document this in [`Doc/library/turtle.rst`](https://github.com/python/cpython/blob/main/Doc/library/turtle.rst).  Be sure to include a `.. versionadded: 3.12` annotation on the new teleport method documentation in there.

**Comment by Fidget-Spinner** [comment]
Terry and Gregory, thank you for taking the time to review this PR.

If you're wondering where to add tests, they are found in `/Lib/test/`. In this case, it is `test_turtle.py`. You can run the test with `./python -m test test_turtle`.  Read the devguide here to learn more https://devguide.python.org/testing/run-write-tests/ .

I also sent you an email to discuss CPython contribution. Please take a look when you have the time. Thanks!

File: Doc/library/turtle.rst
Comment (hugovk) [inline Doc/library/turtle.rst]: The "Docs" CI check is failing with:

```
Error: library/turtle.rst:386: trailing whitespace (trailing-whitespace)
```

https://github.com/python/cpython/actions/runs/4834172590/jobs/8615071055?pr=103974#step:6:16

Let's remove the whitespace at the end:

```suggestion
   with fill_gap=True, which makes the imaginary line traveled during
```

File: Lib/turtle.py
Comment (hugovk) [inline Lib/turtle.py]: Please move 'teleport' down to keep this list alphabetical.

File: Misc/NEWS.d/next/Library/2023-04-28-18-04-23.gh-issue-88773.xXCNJw.rst
Comment (hugovk) [inline Misc/NEWS.d/next/Library/2023-04-28-18-04-23.gh-issue-88773.xXCNJw.rst]: We can also add an automatic link to the turtle docs:

```suggestion
Added :func:`turtle.teleport` to the :mod:`turtle` module to move a turtle to a new point without tracing a line, visible or invisible.  Patch by Liam Gersten.
```

File: Lib/turtle.py
Comment (liam-gersten) [reply Lib/turtle.py]: These changes have now been pushed. Thanks!

**Comment by liam-gersten** [comment]
I have made the requested changes; please review again.

File: Lib/turtle.py
Comment (liam-gersten) [reply Lib/turtle.py]: Done!

**Review by liam-gersten — COMMENTED** [review]
Reviewed

File: Lib/test/test_turtle.py
Comment (gpshead) [inline Lib/test/test_turtle.py:1]: One call in each test case should pass a `fill_gap=` argument as a way to test that the arg is at least accepted.

File: Lib/test/test_turtle.py
Comment (gpshead) [inline Lib/test/test_turtle.py]: assertFalse

**Comment by gvanrossum** [comment]
I wish the turtle module was a 3rd party package, but since it isn't, and is actively used, I think this is a good contribution. I hope @gpshead will come through with his promise of review of the PR. (I didn't see anything objectionable on a quick skim.)

File: Lib/test/test_turtle.py
Comment (liam-gersten) [reply Lib/test/test_turtle.py]: I've changed this in the latest commit, thanks!

File: Lib/test/test_turtle.py
Comment (liam-gersten) [reply Lib/test/test_turtle.py:1]: I've added the argument to the teleport calls in tests.

**Comment by liam-gersten** [comment]
I have made the requested changes; please review again.

**Review by gpshead — APPROVED** [review]

**Pull request merged**
